### PR TITLE
Cond_op with dynamic if-else checked-in

### PR DESCRIPTION
**Pull request by zchen0211**
Implemented the dynamic condition (if/else) op. The result matches my python results. 
TODO: Will implement the backward part soon.

### CODE REVIEW

File: paddle/framework/tensor.h [hunk 1]
```diff
@@ -78,6 +78,9 @@ class Tensor {
   template <typename T>
   inline T* mutable_data(DDim dims, platform::Place place);
 
+  /*! Size of a single element in data() */
+  inline size_t element_size() const { return holder_->element_size(); }
```
Comment (qingqing01) [inline paddle/framework/tensor.h]: Actually, I think there is no need to add this function. A function like `Variable::IsType()` is more useful: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/variable.h#L41

File: paddle/operators/cond_op.cc [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      Tensor* tensor_child = v->GetMutable<Tensor>();
+
+      // Resize child
+      DDim dim = tensor_child->dims();
+      dim[0] = index_[i].size();
+      tensor_child->Resize(dim);
+      tensor_child->mutable_data<float>(dim, platform::CPUPlace());
+
+      Gather<float>(dev_ctx.GetPlace(), tensor_parent, &index_tensors[i],
+                    tensor_child);
```
Comment (qingqing01) [inline paddle/operators/cond_op.cc]: The recurrent operators use explicit type float in Op. `CondOp` has the same problem.  Maybe the `RecurrentOp` and `CondOp` should like:

```
template <T>
class CondOp : public framework::OperatorBase {
}
```

Then specialized the class.

File: paddle/operators/cond_op.cc [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    PADDLE_ENFORCE_EQ(tensor_t_out->dims(), tensor_f_out->dims(),
+                      "Outputs not of the same shape");
+    tensor_out->Resize(tensor_t_out->dims());
+    tensor_out->mutable_data<float>(tensor_out->dims(), platform::CPUPlace());
+  }
+}
+
+void CondOp::Run(const Scope& scope,
+                 const platform::DeviceContext& dev_ctx) const {
+  auto sub_scopes = scope.FindVar("SubScopes")->Get<std::vector<Scope*>>();
```
Comment (Superjom) [inline paddle/operators/cond_op.cc]: enforce scope.FindVar("SubScopes") is not null first
then Get.

File: paddle/operators/cond_op.cc [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  auto index_tensors =
+      scope.FindVar("IndexTensors")->Get<std::vector<Tensor>>();
```
Comment (Superjom) [inline paddle/operators/cond_op.cc]: same as above

File: paddle/operators/cond_op.cc [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      tensor_child->Resize(dim);
+      tensor_child->mutable_data<float>(dim, platform::CPUPlace());
+
+      Gather<float>(dev_ctx.GetPlace(), tensor_parent, &index_tensors[i],
+                    tensor_child);
+    }
+  }
+
+  // Step 3: run
+  for (int i = 0; i < 2; ++i) sub_net_op_[i]->Run(*sub_scopes[i], dev_ctx);
```
Comment (Superjom) [inline paddle/operators/cond_op.cc]: add { }
every cond operator should be wrapped with {}

File: paddle/operators/cond_op.cc [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    AddInput("Cond", "The condition, which is a bool vector");
+    AddInput("Xs", "Inputs of Subnets").AsDuplicable();
+    AddOutput("Outs", "Outputs of Cond_Op after merge").AsDuplicable();
+
+    AddOutput("SubScopes", "sub scopes for true and false branches");
+    AddOutput("IndexTensors", "Index Tensors contains indices for true/false");
+
+    AddComment(R"DOC(
+Sample dependent Cond Operator:
+The equation is: Out[i] = subnet_t[i], if Cond[i] == true
```
Comment (Superjom) [inline paddle/operators/cond_op.cc]: need some indent here

File: paddle/operators/cond_op.h [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  void CreateScope(const framework::Scope& scope) const;
+
+  void CreateIndexTensor(const framework::Scope& scope) const;
+
+  /**
+   * InferShape must be called before Run.
+   */
+  void InferShape(const framework::Scope& scope) const override;
+
+  // Set True Block
```
Comment (Superjom) [inline paddle/operators/cond_op.h]: comment format should be unified.

use 

```
/*
 * some comment
 */
```

File: paddle/operators/cond_op.h [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  void set_truenet(std::unique_ptr<OperatorBase> net) {
+    sub_net_op_[0] = std::move(net);
+  }
+
+  // Set False Block
+  void set_falsenet(std::unique_ptr<OperatorBase> net) {
```
Comment (Superjom) [inline paddle/operators/cond_op.h]: std::unique_ptr<OperatorBase> &&net)

global reference

File: paddle/operators/cond_op.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+#include "glog/logging.h"
+#include "paddle/framework/ddim.h"
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/operator.h"
+#include "paddle/framework/tensor.h"
+#include "paddle/operators/net_op.h"
+
+namespace paddle {
+namespace operators {
+
```
Comment (Superjom) [inline paddle/operators/cond_op.h]: some doc here.

**Pull request merged**